website/content/en/docs/best-practices/observability-best-practices.md

```diff
@@ -65,6 +65,19 @@ The most common types are:
  - `Gauge` Value can be increased and decreased as needed.
 5. `_total` suffix should be used for accumulating count. If your metrics has labels with high cardinality, like `pod`/`container` it usually means that you can aggregate it more, thus it will not require `_total` suffix.
 
+#### Recommended Metrics
+
+* `operator_health_status` - The metric reports the operator health status. It should be reported by the operator itself based on the internal operator validations.
+
+  This metric allows the cluster users to view all operators health with a single metric and improves the overall cluster observability.  
+  Valid health status values are:
+  * `0` - Indicates that the operator is healthy and working as expected.
+  * `1` - Indicates that the operator has some issues that needs to be addressed and can lead to loss of functionality.
+  * `2` - Indicates that the operator is unhealthy and there is a loss of functionality that should be addressed.
+
+  Metric labels:
+  * `name` - The operator name.
```

Comment:
Could you point to precedents for the name and format and/or provide explanations on why they are a good choice?

Quickly looking for similar things, I found:
~~~
name
    kube_node_status_condition
condition
    DiskPressure
    PIDPressure
    MemoryPressure
    Ready
value
    0/1
status
    false/true/unknown
~~~
https://github.com/kubernetes/kube-state-metrics/blob/main/docs/node-metrics.md

~~~
name
    kube_deployment_status_condition
condition
    Available
    Progressing
    ReplicaFailure
value
    0/1
~~~
https://github.com/kubernetes/kube-state-metrics/blob/main/docs/deployment-metrics.md

~~~
name
	cluster_operator_conditions
condition
	Available
	Degraded
	Upgradeable
	Progressing
	Disable
	EvaluationConditionsDetected
	Failing
~~~
https://github.com/openshift/cluster-version-operator/blob/master/install/0000_90_cluster-version-operator_02_servicemonitor.yaml

Reply:
Sure. 

CR Conditions are not the same as statuses, since the list of values of conditions is determined by Kubernetes. For example https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-phase.

I believe we should use numbers to represent the statuses in the values, since string values like: Healthy/Warning/Error, can be implemented in many ways that would cause consistency issues and disagreements, like spelling errors, capital letters and preferences of terminology like : `Critical` instead of `Error`. 
Like you, In Kubernetes, I only found examples where status is reported as a label and the value is a boolean.
That is problematic in our case since it would be very hard to control the implementation of each operator and all this effort would not translate into value for the users. 
We need to keep in mind that kubernetes metrics and most metrics in general, are reported from the same project source, which is much simpler then our case, where the metrics would be reported  from each operator separately. So the metric that we add should be very clear and leave no room for confusion.

I chose 3 statuses to allow some room for reporting cases where there are issues in the operator that should be fixed and are related to its health, but the functionality is still in tact(For example, when the number of replicas is not as it should, but there is still a replica that allows the operator to function). 
Also, The existing health metrics, like the ODF/Ceph/NooBaa health metrics also use this methodology.

Reply:
>  list of values of conditions is determined by Kubernetes.

It is determined by the controller author (here the operator author). Conditions is the mechanism used in Kubernetes and largely followed by most of the extensions for surfacing the status of the intent conveyed by a resource.

> That is problematic in our case since it would be very hard to control the implementation of each operator and all this effort would not translate into value for the users.

I am not sure to follow. The approach taken by authors of node, deployment or cluster operator controllers is to provide an extension point to operator authors through the condition label. Ideally they would use common values: "Available", "Progressing" but could add more if they feel that they have a special need. From the user point of view they would start with common values but could also extend if they feel that the special cases foreseen by the operator authors need their attention.

> We need to keep in mind that kubernetes metrics and most metrics in general, are reported from the same project source,

I would argue that it is not the case for cluster operators.

Reply:
@sradco, thanks for that. I like that it is possible to have a single query for all the operators following the naming pattern. I still think that "conditions" are a natural match for things to monitor but it does not have to happen today. Your PR is already a good improvement and if we feel that a finer granularity based on conditions is needed / get a larger adoption we can always extend our recommendation then.
I am looking forward to this PR getting merged when the other reviewers think it is good to go.

Reply:
The issue is that conditions might not be able to fully indicate the operator health. That is at least the case for our operator. Taking this into account I decided that for now we should focus on the naming and reporting conventions and not the "how to calculate" the metric. I have my thought about that but I think its out of scope for now for this PR. I agree that conditions is defiantly a part of the way to get the operator health status.